blockchain/chaingen/generator.go

```diff
@@ -1021,6 +1082,309 @@ func (g *Generator) CalcNextRequiredDifficulty() uint32 {
+	} else {
+		bn = big.NewInt(int64(mantissa))
+		bn.Lsh(bn, 8*(exponent-3))
+	}
+
+	// Make it negative if the sign bit is set.
+	if isNegative {
+		bn = bn.Neg(bn)
+	}
+
+	return bn
+}
+
+// bigToCompact converts a whole number N to a compact representation using an
+// unsigned 32-bit number.  The compact representation only provides 23 bits of
+// precision, so values larger than (2^23 - 1) only encode the most significant
+// digits of the number.  See compactToBig for details.
+func bigToCompact(n *big.Int) uint32 {
+	// No need to do any work if it's zero.
+	if n.Sign() == 0 {
+		return 0
+	}
+
+	// Since the base for the exponent is 256, the exponent can be treated as
+	// the number of bytes.  So, shift the number right or left accordingly.
+	// This is equivalent to:
+	// mantissa = mantissa / 256^(exponent-3)
+	var mantissa uint32
+	exponent := uint(len(n.Bytes()))
+	if exponent <= 3 {
+		mantissa = uint32(n.Bits()[0])
+		mantissa <<= 8 * (3 - exponent)
+	} else {
+		// Use a copy to avoid modifying the caller's original number.
+		tn := new(big.Int).Set(n)
+		mantissa = uint32(tn.Rsh(tn, 8*(exponent-3)).Bits()[0])
+	}
+
+	// When the mantissa already has the sign bit set, the number is too large
+	// to fit into the available 23-bits, so divide the number by 256 and
+	// increment the exponent accordingly.
+	if mantissa&0x00800000 != 0 {
+		mantissa >>= 8
+		exponent++
+	}
+
+	// Pack the exponent, sign bit, and mantissa into an unsigned 32-bit int and
+	// return it.
+	compact := uint32(exponent<<24) | mantissa
+	if n.Sign() < 0 {
+		compact |= 0x00800000
+	}
+	return compact
+}
+
+// CalcASERTDiff calculates an absolutely scheduled exponentially weighted
+// target difficulty for the given set of parameters using the algorithm defined
+// in DCP0011.
+//
+// The Absolutely Scheduled Exponentially weighted Rising Targets (ASERT)
+// algorithm defines an ideal schedule for block issuance and calculates the
+// difficulty based on how far the most recent block's timestamp is ahead or
+// behind that schedule.
+//
+// The target difficulty is set exponentially such that it is doubled or halved
+// for every multiple of the half life the most recent block is ahead or behind
+// the ideal schedule.
+//
+// The starting difficulty bits parameter is the initial target difficulty all
+// calculations use as a reference.  This value is defined on a per-chain basis.
+// It must be non-zero and less than or equal to the provided proof of work
+// limit or the function will panic.
+//
+// The time delta is the number of seconds that have elapsed between the most
+// recent block and an initial reference timestamp.
+//
+// The height delta is the number of blocks between the most recent block height
+// and an initial reference height.  It must be non-negative or the function
+// will panic.
+//
+// This function is safe for concurrent access.
+func calcASERTDiff(startDiffBits uint32, powLimit *big.Int, targetSecsPerBlock,
```

Comment:
`chaingen` is intended to be entirely independent code from the code that consensus uses since otherwise it would be the equivalent of testing if something is equal to itself and thus wouldn't be good at catching regressions.  You'll notice that `chaingen` doesn't use `standalone` nor `blockchain` (back when it was exported) for anything else either for the same reason.

Also, ultimately the `standalone.CalcASERTDiff` will likely be optimized to avoid using arbitrary precision integers and changed over to use a fixed precision type like `math/uint256` at which point having the separate implementation still based on a big ints will be even more valuable.